### PR TITLE
5249 

### PR DESCRIPTION
**Pull request by cangosta**

### CODE REVIEW

File: CefGlue/Classes.Handlers/CefDownloadHandler.cs [hunk 1]
```diff
@@ -38,13 +38,13 @@ private void on_before_download(cef_download_handler_t* self, cef_browser_t* bro
             CheckSelf(self);
 
             var m_browser = CefBrowser.FromNative(browser);
-            var m_download_item = CefDownloadItem.FromNative(download_item);
-            var m_suggested_name = cef_string_t.ToString(suggested_name);
-            var m_callback = CefBeforeDownloadCallback.FromNative(callback);
+            using (var m_download_item = CefDownloadItem.FromNative(download_item))
+            {
+                var m_suggested_name = cef_string_t.ToString(suggested_name);
+                var m_callback = CefBeforeDownloadCallback.FromNative(callback);
```
Comment (cangosta) [inline CefGlue/Classes.Handlers/CefDownloadHandler.cs:44]: the m_callback is also disposable. Shouldn't we wrap it in a using statement?

File: CefGlue/Classes.Handlers/CefDownloadHandler.cs [hunk 2]
```diff
@@ -64,12 +64,12 @@ private void on_download_updated(cef_download_handler_t* self, cef_browser_t* br
             CheckSelf(self);
 
             var m_browser = CefBrowser.FromNative(browser);
-            var m_download_item = CefDownloadItem.FromNative(download_item);
-            var m_callback = CefDownloadItemCallback.FromNative(callback);
-
-            OnDownloadUpdated(m_browser, m_download_item, m_callback);
+            using (var m_download_item = CefDownloadItem.FromNative(download_item))
+            {
+                var m_callback = CefDownloadItemCallback.FromNative(callback);
```
Comment (cangosta) [inline CefGlue/Classes.Handlers/CefDownloadHandler.cs:69]: the m_callback is also disposable. Shouldn't we wrap it in a using statement?

File: CefGlue/Classes.Handlers/CefRequestHandler.cs [hunk 3]
```diff
@@ -185,7 +185,7 @@ private int on_certificate_error(cef_request_handler_t* self, cef_browser_t* bro
 
             var m_browser = CefBrowser.FromNative(browser);
             var m_request_url = cef_string_t.ToString(request_url);
-            var m_ssl_info = CefSslInfo.FromNative(ssl_info);
+            var m_ssl_info = CefSslInfo.FromNative(ssl_info); // TODO dispose?
             var m_callback = CefCallback.FromNativeOrNull(callback);
 
             var result = OnCertificateError(m_browser, cert_error, m_request_url, m_ssl_info, m_callback);
```
Comment (cangosta) [inline CefGlue/Classes.Handlers/CefRequestHandler.cs:191]: should we wrap the m_callback.Dispose(); below?

File: CefGlue.Interop.Gen/make_interop.py [hunk 4]
```diff
@@ -391,7 +393,7 @@ def make_proxy_g_body(cls):
     result.append(indent + 'if (ptr == null) throw new ArgumentNullException("ptr");')
     result.append(indent + '_self = ptr;')
     if isRefCounted:
-        result.append(indent + 'CefObjectTracker.Track(this);')
+            result.append(indent + 'CefObjectTracker.Track(this);')
```
Comment (joaompneves) [inline CefGlue.Interop.Gen/make_interop.py:396]: fix identation

File: Directory.Build.props [hunk 5]
```diff
@@ -8,6 +8,8 @@
     <PropertyGroup>
         <PackageOutputPath>$(MSBuildProjectDirectory)\..\Nuget\output</PackageOutputPath>
         <Version>91.4472.31</Version>
+<!--        <Version>106.0.26</Version>-->
```
Comment (cangosta) [inline Directory.Build.props:11]: Apply this change

File: CefGlue/Interop/Classes.g/cef_accessibility_handler_t.g.cs [hunk 6]
```diff
@@ -1,4 +1,4 @@
-//
```
[the remark this answers is not in the document]
Comment (cangosta) [reply CefGlue/Interop/Classes.g/cef_accessibility_handler_t.g.cs]: fixed

File: CefGlue.Common/CommonOffscreenBrowserAdapter.cs [hunk 7]
```diff
@@ -40,19 +40,19 @@ private void SendMouseClickEvent(CefMouseEvent mouseEvent, CefMouseButtonType mo
...
+        //         BrowserHost?.SendFocusEvent(true);
+        //     });
+        // }
 
         private void HandleLostFocus()
         {
             WithErrorHandling(nameof(HandleLostFocus), () =>
             { 
-                BrowserHost?.SendFocusEvent(false);
+                BrowserHost?.SetFocus(false);
```
Comment (cangosta) [inline CefGlue.Common/CommonOffscreenBrowserAdapter.cs:55]: Review this change

File: CefGlue.Common/CommonOffscreenBrowserAdapter.cs [hunk 8]
```diff
@@ -40,19 +40,19 @@ private void SendMouseClickEvent(CefMouseEvent mouseEvent, CefMouseButtonType mo
...
         }
 
-        protected override void HandleGotFocus()
-        {
-            WithErrorHandling(nameof(HandleGotFocus), () =>
-            {
-                BrowserHost?.SendFocusEvent(true);
-            });
-        }
+        // protected override void HandleGotFocus()
```
Comment (cangosta) [inline CefGlue.Common/CommonOffscreenBrowserAdapter.cs:43]: Review this change

File: CefGlue/Classes.Handlers/CefRenderProcessHandler.cs [hunk 9]
```diff
@@ -33,19 +33,20 @@ private void on_browser_created(cef_render_process_handler_t* self, cef_browser_
             CheckSelf(self);
 
             var m_browser = CefBrowser.FromNative(browser);
-            var m_extraInfo = CefDictionaryValue.FromNative(extra_info); // TODO dispose?
+            var m_extraInfo = CefDictionaryValue.FromNativeOrNull(extra_info); // TODO dispose?
```
Comment (cangosta) [inline CefGlue/Classes.Handlers/CefRenderProcessHandler.cs:36]: dispose?

File: CefGlue.Common/Helpers/JavascriptToNativeTypeConverter.cs [hunk 10]
```diff
@@ -48,6 +48,14 @@ public static object ConvertToNative(object obj, Type expectedType)
                 return Convert.ToDouble(obj);
             }
 
+            // TEMPORARY FIX for date serialization
```
Comment (cangosta) [inline CefGlue.Common/Helpers/JavascriptToNativeTypeConverter.cs:51]: This needs to be reviewed

File: CefGlue.Common/Helpers/JavascriptToNativeTypeConverter.cs [hunk 11]
```diff
@@ -48,6 +48,14 @@ public static object ConvertToNative(object obj, Type expectedType)
...
+            if (expectedType == typeof(DateTime)
+                && obj is Dictionary<string,object> { Count: 1 } objValue
+                && objValue.TryGetValue("Ticks", out var dateTicks)) {
+                var ticks = Convert.ToDouble(dateTicks);
```
Comment (cangosta) [inline CefGlue.Common/Helpers/JavascriptToNativeTypeConverter.cs:55]: rename to nanoseconds or convert directly to microseconds

File: CefGlue/CefGlue.csproj [hunk 12]
```diff
@@ -1,314 +1,32 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;net45</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0; net45</TargetFrameworks>
```
Comment (cangosta) [inline CefGlue/CefGlue.csproj:4]: Change to net6

File: CefGlue/Classes.Proxies/CefDragData.cs [hunk 13]
```diff
@@ -275,6 +275,12 @@ public void AddFile(string path, string displayName)
             }
         }
 
+        /// <summary>
+        /// Clear list of filenames.
+        /// </summary>
+        public void ClearFilenames()
+            => cef_drag_data_t.clear_filenames(_self);
```
Comment (joaompneves) [inline CefGlue/Classes.Proxies/CefDragData.cs:282]: is this your change or merge result? not following the conventions { }

File: Directory.Build.props [hunk 14]
```diff
@@ -1,13 +1,14 @@
...
-        <CefRedistOSXVersion>91.1.6</CefRedistOSXVersion>
+        <CefRedistVersion>106.0.26</CefRedistVersion>
+        <CefRedistOSXVersion>106.0.26</CefRedistOSXVersion>
     </PropertyGroup>
 
     <PropertyGroup>
         <PackageOutputPath>$(MSBuildProjectDirectory)\..\Nuget\output</PackageOutputPath>
-        <Version>91.4472.31</Version>
+        <Version>106.5249.0</Version>
+
```
Comment (joaompneves) [inline Directory.Build.props:11]:
```suggestion
```

File: CefGlue.Demo.Avalonia/BrowserView.xaml.cs [hunk 15]
```diff
@@ -83,21 +83,21 @@ public async void EvaluateJavascript()
         {
             var result = new StringWriter();
 
-            result.WriteLine(await browser.EvaluateJavaScript<string>("\"Hello World!\""));
+            result.Write(await browser.EvaluateJavaScript<string>("\"Hello World!\""));
```
Comment (joaompneves) [inline CefGlue.Demo.Avalonia/BrowserView.xaml.cs:86]: why this change?

File: CefGlue.Common/Helpers/JavascriptToNativeTypeConverter.cs [hunk 16]
```diff
@@ -11,6 +11,7 @@ internal static class JavascriptToNativeTypeConverter
     {
         private delegate IList ListContructor(int length, out Type genericType);
 
+        private static readonly DateTime _win32Epoch = new DateTime(1601, 1, 1);
```
Comment (joaompneves) [inline CefGlue.Common/Helpers/JavascriptToNativeTypeConverter.cs:14]: why win32?

File: CefGlue/Classes.Handlers/CefRequestHandler.cs (hunk 3 above)
Comment (joaompneves) [reply CefGlue/Classes.Handlers/CefRequestHandler.cs:191]: thats a good question

File: CefGlue.Common/Helpers/JavascriptToNativeTypeConverter.cs (hunk 16 above)
Comment (cangosta) [reply CefGlue.Common/Helpers/JavascriptToNativeTypeConverter.cs:14]: image

File: CefGlue.Demo.Avalonia/BrowserView.xaml.cs (hunk 15 above)
Comment (cangosta) [reply CefGlue.Demo.Avalonia/BrowserView.xaml.cs:86]: The newline added by WriteLine was not working in mac (it wasn't working for 4472 neither)

File: CefGlue/Classes.Proxies/CefDragData.cs (hunk 13 above)
Comment (cangosta) [reply CefGlue/Classes.Proxies/CefDragData.cs:282]: it is the merge result

File: CefGlue.BrowserProcess/Helpers/NativeLibsLoader.cs [hunk 17]
```diff
@@ -12,7 +12,7 @@ internal static class NativeLibsLoader
         /// </summary>
         public static void Install()
         {
-#if NET5_0
+#if NET6_0
```
Comment (joaompneves) [inline CefGlue.BrowserProcess/Helpers/NativeLibsLoader.cs:15]:
```suggestion
#if NET_LATEST
```

File: CefGlue.Interop.Gen/gen-cef3.sh [hunk 18]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+
+# @echo off
+# Determine python path by inspecting registry: HKCU/HKLM SOFTWARE\Python\PythonCore\${PYTHON_VERSION}\InstallPath
+# And then try some default locations: c:\python27amd64 , c:\python27
+/usr/bin/python3 -B cefglue_interop_gen.py --cpp-header-dir include --cefglue-dir ..\CefGlue\ --no-backup
```
[the remark this answers is not in the document]
Comment (cangosta) [reply CefGlue.Interop.Gen/gen-cef3.sh]: done

File: CefGlue.Interop.Gen/gen-cef3.sh [hunk 19]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+
+# @echo off
+# Determine python path by inspecting registry: HKCU/HKLM SOFTWARE\Python\PythonCore\${PYTHON_VERSION}\InstallPath
```
[the remark this answers is not in the document]
Comment (cangosta) [reply CefGlue.Interop.Gen/gen-cef3.sh]: done

File: CefGlue.BrowserProcess/CefGlue.BrowserProcess.csproj [hunk 20]
```diff
@@ -20,11 +20,11 @@
   <ItemGroup>
     <PackageReference Include="System.Runtime.Loader" />
   </ItemGroup>
-    
-  <Target Name="PublishApp" AfterTargets="AfterBuild" Condition="'$(_AssemblyTimestampBeforeCompile)' != '$(_AssemblyTimestampAfterCompile)' and $(TargetFramework.StartsWith('net5')) and '$(IsPublishing)' != 'True'"> 
+
+  <Target Name="PublishApp" AfterTargets="AfterBuild" Condition="'$(_AssemblyTimestampBeforeCompile)' != '$(_AssemblyTimestampAfterCompile)' and $(TargetFramework.StartsWith('net6')) and '$(IsPublishing)' != 'True'">
```
Comment (joaompneves) [inline CefGlue.BrowserProcess/CefGlue.BrowserProcess.csproj:24]:
```suggestion
  <Target Name="PublishApp" AfterTargets="AfterBuild" Condition="'$(_AssemblyTimestampBeforeCompile)' != '$(_AssemblyTimestampAfterCompile)' and $(TargetFramework.StartsWith('$(DotnetVersion)')) and '$(IsPublishing)' != 'True'">
```

File: CefGlue.Demo.Avalonia/BrowserView.xaml.cs (hunk 15 above)
Comment (joaompneves) [reply CefGlue.Demo.Avalonia/BrowserView.xaml.cs:86]: how about using environment.newline?

File: CefGlue.Tests/Javascript/NativeObjectInteropTests.cs [hunk 21]
```diff
@@ -274,11 +274,21 @@ public async Task NativeObjectMethodStringResultIsReturned()
...
 
             var result = await nativeObject.ResultTask;
 
-            Assert.AreEqual(nativeObject.MethodWithDateTimeReturn(), result);
+            // Assert
+            var nativeObjectDateTime = nativeObject.MethodWithDateTimeReturn();
+            var elapsedMicrosecondsFromWin32Epoch =
+                (nativeObjectDateTime - win32Epoch).TotalMilliseconds * 1000;
+
+            Assert.IsTrue(result is IDictionary<string, object>);
```
Comment (joaompneves) [inline CefGlue.Tests/Javascript/NativeObjectInteropTests.cs:290]: I think there's a type assertion

File: CefGlue.Demo.Avalonia/BrowserView.xaml.cs (hunk 15 above)
Comment (cangosta) [reply CefGlue.Demo.Avalonia/BrowserView.xaml.cs:86]: I've already tested that before and we get the same error.

image

File: CefGlue.Tests/Javascript/NativeObjectInteropTests.cs (hunk 21 above)
Comment (cangosta) [reply CefGlue.Tests/Javascript/NativeObjectInteropTests.cs]: done

File: CefGlue.Common/Helpers/JavascriptToNativeTypeConverter.cs (hunk 16 above)
Comment (cangosta) [reply CefGlue.Common/Helpers/JavascriptToNativeTypeConverter.cs]: renamed to windowsEpoch

File: CefGlue/CefGlue.csproj [hunk 22]
```diff
@@ -1,30 +1,32 @@
...
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
-    <LangVersion>8.0</LangVersion>
+    <LangVersion>latest</LangVersion>
   </PropertyGroup>
 
   <ItemGroup>
     <Compile Remove="Wrapper\**" />
     <EmbeddedResource Remove="Wrapper\**" />
     <None Remove="Wrapper\**" />
+    <None Remove="CefGlue.g.props" />
```
[the remark this answers is not in the document]
Comment (cangosta) [reply CefGlue/CefGlue.csproj]: done

File: CefGlue/CefGlue.g.props [hunk 23]
```diff
@@ -0,0 +1,254 @@
+<Project>
```
[the remark this answers is not in the document]
Comment (cangosta) [reply CefGlue/CefGlue.g.props]: done

File: CefGlue.Tests/Events/EventsTests.cs [hunk 24]
```diff
@@ -105,29 +106,118 @@ void OnConsoleMessage(object sender, ConsoleMessageEventArgs e)
...
-            var contextCreatedCompletionSource = new TaskCompletionSource<bool>();
-            var contextReleasedCompletionSource = new TaskCompletionSource<bool>();
+            var contextCreatedEventsCompletionSource = new TaskCompletionSource<bool>();
+
+            void OnJavascriptContextCreated(object sender, JavascriptContextLifetimeEventArgs e)
+            {
+                contextCreatedEventsCompletionSource.SetResult(true);
+            }
+
+            Browser.JavascriptContextCreated += OnJavascriptContextCreated;
```
Comment (joaompneves) [inline CefGlue.Tests/Events/EventsTests.cs:118]: unsubscribe... within a try finally

File: CefGlue.Tests/Events/EventsTests.cs [hunk 25]
```diff
@@ -105,29 +106,118 @@ void OnConsoleMessage(object sender, ConsoleMessageEventArgs e)
...
+
+            await Browser.LoadContent($"<script>1+1</script>");
+            await contextCreatedEventsCompletionSource.Task;
+        }
+
+        [Test]
+        public async Task JavascriptContextCreatedaAreFiredWhenMultipleContextsCreated()
```
Comment (joaompneves) [inline CefGlue.Tests/Events/EventsTests.cs:125]: typo

File: CefGlue.Tests/Events/EventsTests.cs [hunk 26]
```diff
@@ -105,29 +106,118 @@ void OnConsoleMessage(object sender, ConsoleMessageEventArgs e)
...
+                if (e.Frame.IsMain)
+                {
+                    mainFrameContextCreatedEventsCompletionSource.SetResult(true);
+                } else
+                {
+                    innerFrameContextCreatedEventsCompletionSource.SetResult(true);
+                }
+            }
+
+            Browser.JavascriptContextCreated += OnJavascriptContextCreated;
```
Comment (joaompneves) [inline CefGlue.Tests/Events/EventsTests.cs:141]: unsubscribe

File: framework_changes.patch [hunk 27]
```diff
@@ -0,0 +1,252 @@
+Subject: [PATCH] cleaning frameworks
```
Comment (joaompneves) [inline framework_changes.patch:1]: whats this file?

File: Xilium.CefGlue.sln [hunk 28]
```diff
@@ -1,12 +1,10 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio Version 16
-VisualStudioVersion = 16.0.28803.156
+# Visual Studio Version 17
+VisualStudioVersion = 17.4.33122.133
 MinimumVisualStudioVersion = 10.0.40219.1
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "CefGlue", "CefGlue\CefGlue.csproj", "{703FF300-34B9-48D4-A5D6-E741F48D5F95}"
 EndProject
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "CefGlue.WPF", "CefGlue.WPF\CefGlue.WPF.csproj", "{447C6790-E36E-4A54-BFAC-06C92FFF4AA3}"
```
Comment (joaompneves) [inline Xilium.CefGlue.sln:8]: what happened?

File: framework_changes.patch (hunk 27 above)
Comment (cangosta) [reply framework_changes.patch]: It shouldn't have been commited. Removed.

File: CefGlue.Tests/Events/EventsTests.cs (hunk 26 above)
Comment (cangosta) [reply CefGlue.Tests/Events/EventsTests.cs]: done

File: CefGlue.Tests/Events/EventsTests.cs (hunk 25 above)
Comment (cangosta) [reply CefGlue.Tests/Events/EventsTests.cs]: done

File: CefGlue.Tests/Events/EventsTests.cs (hunk 24 above)
Comment (cangosta) [reply CefGlue.Tests/Events/EventsTests.cs]: done

File: Xilium.CefGlue.sln (hunk 28 above)
Comment (cangosta) [reply Xilium.CefGlue.sln:8]: We were having lots of issues with the pipelines. 

@osesantos can give you more details.

Comment (osesantos) [reply Xilium.CefGlue.sln:8]: yes, I wasn't able to build the whole solution, because of dependencies.

Comment (osesantos) [reply Xilium.CefGlue.sln:8]: also, some system dlls used for .net6 lacked some old methods .net472 used, therefore the build of WPF demo was giving us errors because those methods were not available

Comment (joaompneves) [reply Xilium.CefGlue.sln:8]: have you tried .net 6?

Comment (osesantos) [reply Xilium.CefGlue.sln:8]: I didn't, can we use .net6 with WPF? thought we could only use it with .net472

Comment (joaompneves) [reply Xilium.CefGlue.sln:8]: of course... WPF is still on

Comment (osesantos) [reply Xilium.CefGlue.sln:8]: ok, seems fair enough, thanks

Comment (cangosta) [reply Xilium.CefGlue.sln:8]: this is done.

File: Xilium.CefGlue.sln [hunk 29]
```diff
@@ -1,12 +1,10 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio Version 16
-VisualStudioVersion = 16.0.28803.156
+# Visual Studio Version 17
+VisualStudioVersion = 17.4.33122.133
```
Comment (joaompneves) [inline Xilium.CefGlue.sln:4]: mantain this.. revert the other changes in this file

File: CefGlue.BrowserProcess/CefGlue.BrowserProcess.csproj [hunk 30]
```diff
@@ -20,11 +20,11 @@
   <ItemGroup>
     <PackageReference Include="System.Runtime.Loader" />
   </ItemGroup>
-    
-  <Target Name="PublishApp" AfterTargets="AfterBuild" Condition="'$(_AssemblyTimestampBeforeCompile)' != '$(_AssemblyTimestampAfterCompile)' and $(TargetFramework.StartsWith('net5')) and '$(IsPublishing)' != 'True'"> 
+
+  <Target Name="PublishApp" AfterTargets="AfterBuild" Condition="'$(_AssemblyTimestampBeforeCompile)' != '$(_AssemblyTimestampAfterCompile)' and $(TargetFramework.StartsWith('$(DotnetVersion)')) and '$(IsPublishing)' != 'True'">
```
Comment (joaompneves) [inline CefGlue.BrowserProcess/CefGlue.BrowserProcess.csproj:24]:
```suggestion
  <Target Name="PublishApp" AfterTargets="AfterBuild" Condition="'$(_AssemblyTimestampBeforeCompile)' != '$(_AssemblyTimestampAfterCompile)') and '$(IsPublishing)' != 'True'">
```

File: CefGlue.Avalonia/CefGlue.Avalonia.csproj [hunk 31]
```diff
@@ -1,7 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFramework>$(DotnetStandardVersion)</TargetFramework>
```
[the remark this answers is not in the document]
Comment (cangosta) [reply CefGlue.Avalonia/CefGlue.Avalonia.csproj]: done

**Review by joaompneves — APPROVED** [review]
